File: packages/shared-types/src/identity.ts
import { z } from 'zod';

export const ChannelSchema = z.enum(['whatsapp', 'twitter', 'telegram', 'instagram', 'email', 'internal']);
export type Channel = z.infer<typeof ChannelSchema>;

export const PrincipalRoleSchema = z.enum(['end_user', 'business_owner', 'business_team_member', 'business_agent']);
export type PrincipalRole = z.infer<typeof PrincipalRoleSchema>;

export const TenantRoleSchema = z.enum(['owner', 'admin', 'operator', 'viewer']);
export type TenantRole = z.infer<typeof TenantRoleSchema>;

export const TenantStatusSchema = z.enum(['pending', 'active', 'suspended', 'churned']);
export type TenantStatus = z.infer<typeof TenantStatusSchema>;

export const TenantTierSchema = z.enum(['free', 'starter', 'growth', 'pro', 'enterprise']);
export type TenantTier = z.infer<typeof TenantTierSchema>;

export const PrincipalSchema = z.object({
  id: z.string().uuid(),
  display_name: z.string().nullable(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  metadata: z.record(z.unknown()).default({}),
});
export type Principal = z.infer<typeof PrincipalSchema>;

export const IdentifierSchema = z.object({
  id: z.string().uuid(),
  principal_id: z.string().uuid(),
  channel: ChannelSchema,
  identifier: z.string().min(1),
  verified: z.boolean(),
  created_at: z.string().datetime(),
});
export type Identifier = z.infer<typeof IdentifierSchema>;

export const TenantMembershipSchema = z.object({
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  principal_id: z.string().uuid(),
  role: TenantRoleSchema,
  permissions: z.array(z.string()).default([]),
  invited_by: z.string().uuid().nullable(),
  joined_at: z.string().datetime().nullable(),
  created_at: z.string().datetime(),
});
export type TenantMembership = z.infer<typeof TenantMembershipSchema>;

export const LinkingCodeSchema = z.object({
  code: z.string().min(4).max(32),
  principal_id: z.string().uuid(),
  source_channel: ChannelSchema,
  expires_at: z.string().datetime(),
  used_at: z.string().datetime().nullable(),
  created_at: z.string().datetime(),
});
export type LinkingCode = z.infer<typeof LinkingCodeSchema>;

// E.164 phone number normalisation.
const E164 = /^\+[1-9]\d{6,14}$/;

export function normalizeIdentifier(channel: Channel, raw: string): string {
  if (channel === 'whatsapp') {
    const digits = raw.replace(/[^\d+]/g, '');
    const candidate = digits.startsWith('+') ? digits : `+${digits}`;
    if (!E164.test(candidate)) {
      throw new Error(`invalid E.164 phone number: ${raw}`);
    }
    return candidate;
  }
  if (channel === 'twitter') {
    return raw.replace(/^@/, '').toLowerCase();
  }
  if (channel === 'email') {
    return raw.trim().toLowerCase();
  }
  return raw;
}

File: packages/python-shared/veda_shared/schemas/identity.py
"""Identity & tenant primitives. Pydantic mirrors of TS identity.ts."""

from __future__ import annotations

import re
from datetime import datetime
from enum import Enum
from typing import Any

import phonenumbers
from pydantic import BaseModel, ConfigDict, Field


class Channel(str, Enum):
    WHATSAPP = "whatsapp"
    TWITTER = "twitter"
    TELEGRAM = "telegram"
    INSTAGRAM = "instagram"
    EMAIL = "email"
    INTERNAL = "internal"


class PrincipalRole(str, Enum):
    END_USER = "end_user"
    BUSINESS_OWNER = "business_owner"
    BUSINESS_TEAM_MEMBER = "business_team_member"
    BUSINESS_AGENT = "business_agent"


class TenantRole(str, Enum):
    OWNER = "owner"
    ADMIN = "admin"
    OPERATOR = "operator"
    VIEWER = "viewer"


class TenantStatus(str, Enum):
    PENDING = "pending"
    ACTIVE = "active"
    SUSPENDED = "suspended"
    CHURNED = "churned"


class TenantTier(str, Enum):
    FREE = "free"
    STARTER = "starter"
    GROWTH = "growth"
    PRO = "pro"
    ENTERPRISE = "enterprise"


class Principal(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    display_name: str | None = None
    created_at: datetime
    updated_at: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)


class Identifier(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    principal_id: str
    channel: Channel
    identifier: str
    verified: bool
    created_at: datetime


class TenantMembership(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    tenant_id: str
    principal_id: str
    role: TenantRole
    permissions: list[str] = Field(default_factory=list)
    invited_by: str | None = None
    joined_at: datetime | None = None
    created_at: datetime


class LinkingCode(BaseModel):
    model_config = ConfigDict(extra="ignore")

    code: str
    principal_id: str
    source_channel: Channel
    expires_at: datetime
    used_at: datetime | None = None
    created_at: datetime


_TWITTER_HANDLE = re.compile(r"^[A-Za-z0-9_]{1,15}$")


def normalize_identifier(channel: Channel | str, raw: str) -> str:
    """Match TS shared-types/src/identity.ts:normalizeIdentifier semantics."""
    ch = channel.value if isinstance(channel, Channel) else channel
    if ch == Channel.WHATSAPP.value:
        # Try to parse as phone number — accept any country, output E.164.
        try:
            parsed = phonenumbers.parse(raw, "IN")
        except phonenumbers.NumberParseException as e:
            raise ValueError(f"invalid phone number: {raw}") from e
        if not phonenumbers.is_valid_number(parsed):
            raise ValueError(f"invalid phone number: {raw}")
        return phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)
    if ch == Channel.TWITTER.value:
        candidate = raw.lstrip("@").lower()
        if not _TWITTER_HANDLE.match(candidate):
            raise ValueError(f"invalid twitter handle: {raw}")
        return candidate
    if ch == Channel.EMAIL.value:
        return raw.strip().lower()
    return raw
